### DISCUSSION TITLE
Creating a tree sequence object within a C API integrated program

### DISCUSSION THREAD

**Discussion by mawassw** [tskit-dev/tskit#2726] [Q&A] [answered]
Hi everyone,

My question pertains to creating a valid tree sequence object from a table collection object within a C-based forward in time evolution simulation program intermittently, before the simplification process, to analyze the trees while they contain all genealogical information. 

The issue I am facing is in the following sequence of pseudocode:

**table collection object → sort table collection object using _tsk_table_collection_sort()_ → rebuild index of table collection using _tsk_table_collection_build_index()_ → initialize tree sequence object with the table collection object.** 

Is this sequence of pseudocode correct? I understand that I have to sort the table collection object to get a valid tree sequence but sorting drops the index of the table collection object and I can't create a tree sequence object with an unindexed table collection object. Hence, I am building the index and then initializing the tree sequence object.

NOTE: These steps are working without any error but the tree sequence object I obtain has 0 num_samples which does not make sense. Is the rebuilding of the index possibly leading to this issue? 

Thanks for the help! I can provide more information if needed.

**Comment by benjeffery** [comment, marked as answer]
On first glance this seems right.

Are any of your nodes marked as sample nodes? (i.e. they have the `TSK_NODE_IS_SAMPLE` in their `flags`). It's perfectly valid to have a tree sequence with zero samples as they are just "special" nodes. See https://tskit.dev/tskit/docs/stable/data-model.html#node-table

**Reply by jeromekelleher** [reply]
If you could print out the values (using e.g. tsk_table_collection_print_state(tables, stout) ) for a tiny example it would help to debug.

**Reply by mawassw** [reply]
Hello Ben and Jerome, 

I printed out the state of the tables and I found out the issue. It is as Ben pointed out:
The flag column in the node table of the table collection object shows that all nodes have a 0 meaning not in sample. In our program, we use an array to track the nodes alive in the population and when we simplify the table collection object, we use this node array to inform the function doing the simplification.
I think what I need to do is update the flag of the node rows that are present in the node array as 1 to be present in the sample which will allow me to use the created tree sequence based on the table collection object (every X generations) to do the analysis I need (calculate time of fixation of fixed mutations) iteratively on the trees. Then when I am done I reupdate the rows back to 0 and then do the simplification process on the table collection object the usual way.

Any suggestions on this solution?

**Reply by benjeffery** [reply]
I think what you say seems a good solution - note that you can use the `update_sample_flags` in `simplify` to make the nodes that you specify get flagged as sample nodes.

**Reply by jeromekelleher** [reply]
I thought the samples should be flagged automatically by simplify though... Can you show us a snippet of your code where you're doing this?

I assume you've seen the haploid wright fisher example simulator in the documentation?

**Reply by benjeffery** [reply]
> I thought the samples should be flagged automatically by simplify though... Can you show us a snippet of your code where you're doing this?

You are correct, the default is `True`.

**Reply by mawassw** [reply]
I'll share some code snippets to help clarify things (sorry for the long message).

Re simplification 
we use the default way of simplifying 
_returnvaluefortskfunctions = tsk_table_collection_simplify(&treesequencetablecollection, wholepopulationnodesarray, (2*maxPopSize), 0, NULL);_

after this simplification, we reset the indexes of the node array as such:
_for (k = 0; k < (2*maxPopSize); k++) {
                            wholepopulationnodesarray[k] = k;
                        }_

Re flagging nodes
First, our model is based on non-WF model. When we add a row to the node table when a new individual is born, we don't flag that they are part of the sample. We add them to the _wholepopulationnodesarray_ which we use to inform the simplification function. The reason for this, is that initially our goal was to assess the final sample of our runs. But now, since I am intermittently (during a single run) creating a tree sequence from a table collection as such every X generations (followed by simplification after the iterative tree analysis is done):

//sort the table
 _returnvaluefortskfunctions = tsk_table_collection_sort(&treesequencetablecollection, NULL, 0);
 check_tsk_error(returnvaluefortskfunctions);
//reindex the table
returnvaluefortskfunctions = tsk_table_collection_build_index(&treesequencetablecollection, 0);
check_tsk_error(returnvaluefortskfunctions);
 //create tree sequence object
tsk_treeseq_t treesequence;
returnvaluefortskfunctions = tsk_treeseq_init(&treesequence, &treesequencetablecollection, 0);
check_tsk_error(returnvaluefortskfunctions);_

The issue I was getting was that the created tree sequence I was obtaining had 0 samples. So when I am iteratively analyzing its trees, and use the _tsk_treeseq_get_samples(tree->tree_sequence);_ my analysis broke down. The solution I am testing right now is just the use _wholepopulationnodesarray_ as the sample node to find things like the MRCA of the sample and find fixed mutations in the sample, as well as as attempt to get the time of fixation with our own algorithm.

I can provide more details if necessary.

**Reply by jeromekelleher** [reply]
> But now, since I am intermittently (during a single run) creating a tree sequence from a table collection as such every X generations (followed by simplification after the iterative tree analysis is done):

Ah, you're creating the ts *before* simplification. That all makes sense then - you just need to set the TSK_NODE_IS_SAMPLE flag for the nodes in the current population before you call ``tsk_treeseq_init`` and it should all work fine.

**Reply by mawassw** [reply]
Thanks Jerome, I actually came to this conclusion recently. Having you reach the same thing is reassuring. Just for clarity on how to go about this: I just use _tsk_node_table_update_row_ function to update the is_sample flag of the nodes that are in the population, correct?
and for the simplification process, I don't need to reset them back since the _tsk_table_collection_simplify_ function unsets it when we provide a sample to the function, correct?

**Reply by jeromekelleher** [reply]
You can actually update the flags in the columns directly if you prefer - it's a bit simpler and more efficient. Something like
```c
for (j = 0; j < population_size; j++) {
     tables.nodes->flags[population[j]] = TSK_NODE_IS_SAMPLE;
}
```

But update_row is equivalent.

Simplify will reset the sample flags, correct.
